### ISSUE TITLE
[Mobile Payments] Refactor WC_Payment_Gateway_WCPay process_payment and process_payment_for_order to allow re-use by an endpoint

### ISSUE THREAD

**Issue by allendav** [Automattic/woocommerce-payments#1514]
To avoid duplicating (and keeping in sync) the logic between the plugin and the mobile apps, instead of having the mobile apps replicate the order logic in `process_payment` and `process_payment_for_order` in , it would be better if `process_payment` and `process_payment_for_order` in `includes/class-wc-payment-gateway-wcpay.php` were refactored in such a way to allow the mobile app to trigger the same logic via an endpoint.

cc @ctarda  @malinajirka

**Comment by jrodger** [comment]
If I'm remembering correctly, this was one of the use cases @v18 had in mind when `process_payment_for_order` was refactored out of `process_payment`.

The idea was to create a method with a known set of inputs (as opposed to `process_payment` that requires all sorts of global state) that could be unit tested or called from other contexts. I appreciate there is also logic in `process_payment` that the mobile app would like to avoid duplicating, but hopefully `process_payment_for_order` is a good start, and I'd love if we could follow the same sort of philosophy when abstracting out anything else required from `process_payment`.

**Comment by kalessil** [comment]
@elazzabi: can you look into this please:
- ist it relevant at yet
- how many lines of duplicate code is this

Based on your estimations, we;ll see how to proceed here.

**Comment by elazzabi** [comment]
After taking a look, here's what is still in `process_payment`:

### Fraud prevention services
This gets the fraud prevention service, double-checking the request is good to go, then executes it. I'm not sure if the fraud prevention service is relevant to the mobile/IPP work.

### Rate limiting
This checks if there is a rate limit before executing the payment. Again, not sure if this is still valid in the case of mobile payments, but can be easily refactored.

### Error handling
The payment is wrapped in a try/catch, and a lot of if/else is used to handle the error raised.

Only two things can be refactored here I believe 1. Marking the payment as `failed` in a particular case (changing the subscription payment info - again probably not an IPP thing) 2. Bumping the rate limiter.

All other error handlers seem to be injecting some error messages with HTML, not sure if that's something mobile can render/use.

-----
So to sum up, I don't believe there is much to refactor here really. But open to other suggestions

**Comment by kalessil** [comment]
Thanks for looking into @elazzabi, let's close this issue then.
